docsource/use_cases/field_fast_precomputation.rst

```diff
@@ -0,0 +1,149 @@
+            comodel_name="product.attribute.value",
+            relation="product_attribute_value_product_template_attribute_line_rel"
+        )
+        value_count = fields.Integer(
+            compute="_compute_value_count",
+            store=True,
+        )
+
+    @api.depends("value_ids")
+    def _compute_value_count(self):
+        for record in self:
+            record.value_count = len(record.value_ids)
+
+See `Full v15 Code Source <https://github.com/odoo/odoo/blob/2f817a7b36cc7e5ab235829eecd61a1d71ce546e/addons/product/models/product_attribute.py#LL189-L209>`_.
+
+Analysis
+--------
+
+.. code-block:: text
+
+    ---Fields in module 'product'---
+
+    product      / product.template.attribute.line / value_count (integer):
+        NEW isfunction: function, stored
+
+See `Full v15 Analysis File <https://github.com/OCA/OpenUpgrade/blob/97491cb7d9a8ed494a49cf1db9b7fc8852aac254/openupgrade_scripts/scripts/product/15.0.1.2/upgrade_analysis.txt#L16>`_.
+
+
+
+Result without migration script / Expected Result
+-------------------------------------------------
+
+The new field will be computed, by the ORM.
+So in absolute terms, there is nothing to do.
+However, the calculation will be performed for all the rows of the table.
+If the table contains a very large amount of data, the calculation can be slow:
+Indeed, even if the ORM contains some prefetch system for reading the data,
+the writing will be done item by item, via an SQL request of the UPDATE type,
+which is not at all optimized, and will generate a (quasi) linear type of time complexity.
+
+See : https://en.wikipedia.org/wiki/Time_complexity#Linear_time
+
+As a result, some compute on huge database can take hours, or days, or worse,
+generate an insufficient memory error, which stops the migration.
+
+These inconveniences obviously depend on the size of the table in question
+and the power and the configuration of the server that is performing the migration.
+
+Contribution to OpenUpgrade
+---------------------------
+
+Update ``upgrade_analysis_work.txt`` file
+"""""""""""""""""""""""""""""""""""""""""
+
+Add a comment after the line:
+
+.. code-block:: text
+
+    product      / product.template.attribute.line / value_count (integer):
+        NEW isfunction: function, stored
+    # DONE: pre-migration: fast computed value_count
+
+See `Full v15 Work Analysis File <https://github.com/OCA/OpenUpgrade/blob/97491cb7d9a8ed494a49cf1db9b7fc8852aac254/openupgrade_scripts/scripts/product/15.0.1.2/upgrade_analysis_work.txt#LL32-L33>`_.
+
+Write migration Script
+""""""""""""""""""""""
+
+in the ``pre-migration.py`` script add:
+
+.. code-block:: python
+
+    from openupgradelib import openupgrade
+
+    def compute_product_template_attribute_line_value_count(env):
+        openupgrade.logged_query(
+            env.cr,
+            """
+            ALTER TABLE product_template_attribute_line
+            ADD IF NOT EXISTS COLUMN value_count integer
+            """,
+        )
```

Comment:
Isn't it easier to use the [`add_fields()`](https://github.com/OCA/openupgradelib/blob/master/openupgradelib/openupgrade.py#L2739) function from `openupgradelib` instead of adding the column by a custom SQL request ?

Reply:
hehe ! you're right !

Reply:
done.